src/lib/hash/blake2s/blake2s.cpp

```diff
@@ -0,0 +1,208 @@
+/*
+ * BLAKE2s
+ * (C) 2023           Richard Huveneers
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+
+#include <botan/internal/blake2s.h>
+
+#include <botan/exceptn.h>
+#include <botan/internal/fmt.h>
+#include <botan/internal/loadstor.h>
+#include <botan/internal/rotate.h>
+
+namespace Botan {
+
+namespace {
+
+// RFC7693 - copy begin
+
+// Initialization Vector.
+
+const uint32_t blake2s_iv[8] = {
+   0x6A09E667, 0xBB67AE85, 0x3C6EF372, 0xA54FF53A, 0x510E527F, 0x9B05688C, 0x1F83D9AB, 0x5BE0CD19};
+
+// Mixing function G.
+
+inline void B2S_G(uint8_t a, uint8_t b, uint8_t c, uint8_t d, uint32_t x, uint32_t y, uint32_t* v) {
+   v[a] = v[a] + v[b] + x;
+   v[d] = rotr<16>(v[d] ^ v[a]);
+   v[c] = v[c] + v[d];
+   v[b] = rotr<12>(v[b] ^ v[c]);
+   v[a] = v[a] + v[b] + y;
+   v[d] = rotr<8>(v[d] ^ v[a]);
+   v[c] = v[c] + v[d];
+   v[b] = rotr<7>(v[b] ^ v[c]);
+}
+
+// Compression function. "last" flag indicates last block.
+
+void blake2s_compress(blake2s_ctx& ctx, bool last) {
+   const uint8_t sigma[10][16] = {{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15},
+                                  {14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3},
+                                  {11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4},
+                                  {7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8},
+                                  {9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13},
+                                  {2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9},
+                                  {12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11},
+                                  {13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10},
+                                  {6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5},
+                                  {10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0}};
+   int i;
```

Comment:
Instead of global `i` use an index scoped to the loops. Also use `size_t` for loop iteration variables of this kind unless the signedness properties are actually used.